### DISCUSSION TITLE
Initial Residuals and Convergence in transient

### DISCUSSION THREAD

**Discussion by Traiwit** [idaholab/moose#28364] [Q&A General]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Hi everyone,

I'm running into a weird issue with my transient analysis in MOOSE. In each analysis, the first frame (like F001) starts off with a massive initial residual around 1e+13 and converges pretty smoothly. But then, the next frame (F002) starts much lower, around 1e+08, and just won't converge.

Even weirder, if I kick off a brand new simulation using the end results of F001 as the starting point via IC, the first step (now F002) starts high again at 1e+13 and converges no problem. But the next one (F003)? Back to the same old problem. This keeps happening.

So, here's what I'm wondering:

1. Why does the first frame always start with a high residual when it’s the first in an analysis but not when it's a subsequent frame?
2. Any idea why this is happening or how to keep the residuals high like in the first frame for all steps?

Would really appreciate any thoughts or tips on this. 

Thank you!

**Comment by GiudGiud** [comment]
Hello

are you using automatic_scaling by any chance?

**Reply by GiudGiud** [reply]
those scaling factors are determined on the first step, and could be wildly wrong for the second step (with a lot of bad luck, this may be far fetched)

**Reply by Traiwit** [reply]
Hi @GiudGiud, no automatic_scaling is applied

my simulation always starts very 1e13 at the very first step, which is good for convergence (via nl_rel_tol)

**Reply by GiudGiud** [reply]
> which is good for convergence (via nl_rel_tol)

well that's kind of an artefact though? Does it get the residual as low as it gets it to in step2?
If you do step 1: 1e13 -> 1e5 (and cant go any lower) ->converged due to rel_tol
step 2 1e8 -> 1e5 (and cant go any lower) -> not converged 
that would mean step 1 has the same problem as step2

**Reply by Traiwit** [reply]
this is a very fair point

relaxing abs_nl_tol is probs one of the ways to fix this then

**Reply by GiudGiud** [reply]
assuming you still like the solution though.

I find that scaling helps with these things. Either automatic or manual.
How many equations (variables) are you solving?

**Reply by Traiwit** [reply]
just 1 variable (pore water pressure) with 2 kernels (`PorousFlowMassTimeDerivative` and `PorousFlowAdvectiveFlux`), dont think scaling would help at all in this case.

I feel like my only ways out are either: make the system easier to solve like increase excavated permeability, relaxing capillary pressure, etc (which might not be not very physically accurate) OR relaxing `abs_nl_tol`

**Reply by Traiwit** [reply]
Hi @GiudGiud I don't want to create a new thread, but just wondering, is there any userobject or any class that I can modify value of some nodes variable before the system is being solved?

my idea is: at timestep_begin, i let pwp (variable) of excavated elements become 0 and start solving from there?

**Reply by GiudGiud** [reply]
only before? If so then initial condition objects are meant for that.

**Reply by GiudGiud** [reply]
oh on timestep_begin. Well we are introducing the Correctors system to do this this week.
there's one example, the PointwiseRenormalizeVector

**Reply by Traiwit** [reply]
oh nice, sounds interesting, does `the Correctors system` do similar to what i want?

**Reply by GiudGiud** [reply]
the object I mentioned already exists btw, it s a user object
Correctors will be meant to modoify the nonlinear solution

**Reply by Traiwit** [reply]
![image](https://github.com/user-attachments/assets/98ff0c8b-fb4a-4b99-9162-c85df6785903)

I couldnt find anything, or you meant i should modify from `PointwiseRenormalizeVector` ?

**Reply by GiudGiud** [reply]
Correctors is in a PR now.
in userobjects you should find PointwiseRenormalizeVector which is a fine example of modifying a solutionvector

another example is the NSPRessurePin UO in navier stokes

**Reply by Traiwit** [reply]
Hi @GiudGiud 

here is my updated version of Corrector:

a quick idea is: let pwp (variable) of element with `state == 1` equal to `target_pwp`

```
//* This file is part of the MOOSE framework
//* https://www.mooseframework.org
//*
//* All rights reserved, see COPYRIGHT for full restrictions
//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
//*
//* Licensed under LGPL 2.1, please see LICENSE for details
//* https://www.gnu.org/licenses/lgpl-2.1.html

#include "ExcavPwp.h"
#include "MooseError.h"
#include "NonlinearSystemBase.h"

#include "libmesh/numeric_vector.h"
#include "libmesh/int_range.h"

registerMooseObject("moose_behApp", ExcavPwp);

InputParameters
ExcavPwp::validParams()
{
  InputParameters params = GeneralUserObject::validParams();
  params.addClassDescription(
      "Pointwise renormalize the solution of a set of variables comprising a vector");
  params.addCoupledVar("variable", "Variables comprising the vector");
  params.addParam<Real>("target_pwp", 0.0, "Desired norm for the coupled variable vector");
  params.addParam<MaterialPropertyName>("state",  0,   "state");

  return params;
}

ExcavPwp::ExcavPwp(const InputParameters & parameters)
  : GeneralUserObject(parameters),
    _mesh(_fe_problem.mesh()),
    _var_names(getParam<std::vector<VariableName>>("variable")),
    _state(getMaterialProperty<Real>("state")),
    _target_pwp(getParam<Real>("target_pwp"))
{
  const MooseVariableFieldBase * first_var = nullptr;
  for (const auto & var_name : _var_names)
  {
    auto & var = _fe_problem.getVariable(0, var_name);
    if (!first_var)
      first_var = &var;
    else
    {
      // check order and family for consistency
      if (first_var->feType() != var.feType())
        paramError("v", "All supplied variables must be of the same order and family.");
      // check block restriction for consistency
      if (!first_var->hasBlocks(var.blockIDs()) || !var.hasBlocks(first_var->blockIDs()))
        paramError("v", "All supplied variables must have the same block restriction.");
    }

    if (_sys.number() != var.sys().system().number())
      paramError("v", "Variables must be all in the non-linear system.");

    if (var.isArray())
    {
      const auto & array_var = _fe_problem.getArrayVariable(0, var_name);
      for (unsigned int p = 0; p < var.count(); ++p)
        _var_numbers.push_back(_sys.system().variable_number(array_var.componentName(p)));
    }
    else
      _var_numbers.push_back(_sys.system().variable_number(var_name));
  }
}

void
ExcavPwp::initialize()
{
  // do one solution.close to get updated
  _sys.system().solution->close();
}

void
ExcavPwp::execute()
{
  auto & dof_map = _sys.system().get_dof_map();
  const auto local_dof_begin = dof_map.first_dof();
  const auto local_dof_end = dof_map.end_dof();

  std::vector<std::vector<dof_id_type>> dof_indices(_var_numbers.size());
  std::vector<Real> cache(_var_numbers.size());

  for (const auto & elem : *_mesh.getActiveLocalElementRange())
  {
    // Retrieve the state value for the current element
    const Real & state_value = _state[elem->id()]; // Access state using the element ID

    if (state_value == 1.0)
    {
      // Prepare variable DOFs
      for (const auto i : index_range(_var_numbers))
      {
        dof_map.dof_indices(elem, dof_indices[i], _var_numbers[i]);

        // Check that all vars have the same number of DOFs
        mooseAssert(dof_indices[i].size() == dof_indices[0].size(),
                    "All specified variables should have the same number of DOFs");
      }

      // Get the current solution
      auto & solution = _sys.solution();

      // Loop over all DOFs
      for (const auto j : index_range(dof_indices[0]))
      {
        // Check if the first variable's DOFs are local (if they are, all other variables should
        // have local DOFS as well)
        if (dof_indices[0][j] > local_dof_end || dof_indices[0][j] < local_dof_begin)
          continue;

        // Set the variable to target_pwp
        for (const auto i : index_range(_var_numbers))
          solution.set(dof_indices[i][j], _target_pwp);
      }
    }
  }
}

void
ExcavPwp::finalize()
{
  for (const auto s : make_range(3))
    if (_sys.hasSolutionState(s))
      _sys.solutionState(s).close();

  _sys.update();
}
```

it seems like the updated results aren't matched with STATE == 1, i'm not so sure where does the mismatch is from.

![image](https://github.com/user-attachments/assets/fba06d63-217d-4f5d-9cf3-d1b06a05cfc7)

![image](https://github.com/user-attachments/assets/9b19bd03-d728-47dc-a958-602d2c02cd94)

**Reply by Traiwit** [reply]
Hi @GiudGiud 

```
void
ExcavPwp::execute()
{
  _console << "DEBUG: ExcavPwp:: UPDATING EXCAV PWP" << std::endl;
  auto & dof_map = _sys.system().get_dof_map();
//   const auto local_dof_begin = dof_map.first_dof();
//   const auto local_dof_end = dof_map.end_dof();

  const auto & mesh = _mesh.getMesh();
  const unsigned int n_elem = mesh.n_elem();

  std::vector<dof_id_type> dof_indices;

    //   for (const auto & elem : *_mesh.getActiveLocalElementRange())
  for (unsigned int i = 0; i < n_elem; ++i)
  {
    // Retrieve the state value for the current element
    // const Real & state_value = _state[elem->id()]; // Access state using the element

    // Real state_value = _read_prop_user_object->getData(mesh.elem_ptr(elem_id), _csv_col_index);
    const Elem * elem = mesh.elem_ptr(i);
    Real state_value = _read_prop_user_object->getData(elem, _csv_col_index);

    if (std::abs(state_value - 1.0) < 1e-6) // Ensure the state is exactly 1.0
    {
      // Prepare DOF indices for the single variable
      dof_map.dof_indices(elem, dof_indices, _var_numbers[0]);

      // Get the current solution
      auto & solution = _sys.solution();

      // Loop over all DOFs for the current element
      for (const auto dof_index : dof_indices)
      {

        // Set the variable to target_pwp
        solution.set(dof_index, _target_pwp);
      }
    }
  }
}
```

I got it to work, for some reason _state via material property does work BUT _state directly from CSV works

I have a question, can i treat is as an IC if I run it at `TIMESTEP_BEGIN`?

**Reply by GiudGiud** [reply]
could you run it on INITIAL?

timestep_begin is close to an IC, but you still have to check with Debug/show_execution_order=ALWAYS to see what got executed before this UO was run.

**Reply by Traiwit** [reply]
Hi @GiudGiud 

I mean I want to project this before it even starts solving at every timestep, i found one thing strange

at the beginning of timestep, it does what I want, projecting 0 pore pressure at excavated elements (drain water out)

![image](https://github.com/user-attachments/assets/baa2ad44-2219-4ef8-bd8c-0bfe8c0b6389)

but then after first linear iteration (I saved .e at every nonlinear), that area got filled with water again, and trying to drain water out via BC

basically it just ignore the 0 pore pressure what I projected at timestep_begin

![image](https://github.com/user-attachments/assets/670fd46e-3998-4bc5-ac7d-3e32e1d4bd6b)

![image](https://github.com/user-attachments/assets/763e6882-0b24-4e0d-99f5-ee909f7bdf63)

Question is: How to make it just take what I projected and start solving from there at every timestep_begin?

Kind regards,
Traiwit

**Reply by GiudGiud** [reply]
interesting. Basically you want to follow what the ICs do. You may be modifying the wrong vector wrt the solve, or are missing a "solution.update" call in this user object.
This is run in parallel too right?

**Reply by Traiwit** [reply]
Hi @GiudGiud i'm back from a month holiday, first day back and I got a question for you :)

```
void
ExcavPwp::execute()
{
  _console << "DEBUG: ExcavPwp:: UPDATING EXCAV PWP" << std::endl;
  auto & dof_map = _sys.system().get_dof_map();

  const auto & mesh = _mesh.getMesh();
  const unsigned int n_elem = mesh.n_elem();

  std::vector<dof_id_type> dof_indices;

  // Get access to the solution vector
  NumericVector<Number> & solution = *_sys.system().solution;

  for (unsigned int i = 0; i < n_elem; ++i)
  {
    const Elem * elem = mesh.elem_ptr(i);
    Real state_value = _read_prop_user_object->getData(elem, _csv_col_index);

    if (state_value > 0 && state_value <= 1) // Ensure the state is exactly 1.0
    {
      // Prepare DOF indices for the single variable
      dof_map.dof_indices(elem, dof_indices, _var_numbers[0]);

      // Loop over all DOFs for the current element
      for (const auto dof_index : dof_indices)
      {
        // Get the current solution value at the DOF
        Number current_value = solution.get(dof_index);

        // If the current value is positive, set it to _target_pwp
        if (current_value > 0.0)
        {
          solution.set(dof_index, _target_pwp);
        }
        // If the value is negative, do nothing (it remains unchanged)
      }
    }
  }

  // Finalize the solution vector after modifications
  solution.close();
}

```

so this is the current version of the code, I solved the previous question I asked you, now I want to have a new condition for this object, such that:

when STATE == 1, if the solution is already negative, leave it as it, if it's positive make it equal to `_target_pwp`

which is this part of the .C

```
 if (state_value > 0 && state_value <= 1) // Ensure the state is exactly 1.0
    {
      // Prepare DOF indices for the single variable
      dof_map.dof_indices(elem, dof_indices, _var_numbers[0]);

      // Loop over all DOFs for the current element
      for (const auto dof_index : dof_indices)
      {
        // Get the current solution value at the DOF
        Number current_value = solution.get(dof_index);

        // If the current value is positive, set it to _target_pwp
        if (current_value > 0.0)
        {
          solution.set(dof_index, _target_pwp);
        }
        // If the value is negative, do nothing (it remains unchanged)
      }
    }
```


i'm having a problem with getting the current solution, I tried 

 `Number current_value = solution.get(dof_index);`

but it doesn't compile, is there another way to obtain the node value?

Thank you!

**Reply by GiudGiud** [reply]
get() is for multiple dofs at once
https://mooseframework.inl.gov/docs/doxygen/libmesh/classlibMesh_1_1NumericVector.html#a0a56cd3b94566f25b4cf3e18a7440a08
the () (parenthesis operator) should be the one for a single dof
https://mooseframework.inl.gov/docs/doxygen/libmesh/classlibMesh_1_1NumericVector.html#a28edc288b838504336b7276f04a3a3a9

**Reply by Traiwit** [reply]
Hi @GiudGiud 

now it compiles, but i got `Segmentation fault` error from this line (I have debug/isolate the code, i'm sure it's this specific line)

        `Real current_value = solution(dof_index);`

not sure why, maybe there is another ay to grab solution from a node?

**Reply by GiudGiud** [reply]
then you must be accessing out of bounds. 
Is this in parallel?

> maybe there is another ay to grab solution from a node?

yes actually, accessing the solution vector is not the sanctioned way. 
You should use either the coupleable  interface's routines (see coupleable.h), or variable.getNodalValue() from the mooseVariableFE.h

**Reply by Traiwit** [reply]
yes it's in parallel

if i use `getNodalValue()`, how do i update the value like the one below:

`solution.set(dof_index, _target_pwp);`

**Reply by GiudGiud** [reply]
you can try this routine on the variable
```
  virtual void setNodalValue(const OutputType & value, unsigned int idx = 0) override;
```
Are you developing in opt mode?
You should hit an assert on an OOB if you are developing in devel (highly recommended) or debug (when things go south)
To change the compilation method, you can use `METHOD=devel make -j6`
or `export METHOD=devel`

**Reply by Traiwit** [reply]
Noted, Thanks @GiudGiud!

hmm I didnt know about developing in opt mode at all, dont even know what it does.

btw, is there any example object that I can have a look for `setNodalValue`? 

it's a bit late for me now, I will have a look at them first thing tomorrow.

**Reply by GiudGiud** [reply]
in opt mode a lot of simple checks are removed for performance purposes.
in debug a lot of non-trivial poorly-scaling checks are added, degrading performance a lot
devel is a healthy middle ground, I rarely use the other modes

you can see setNodalValue being used in AuxKernel.C and in InitalConditonTempl.C
